ckan/logic/action/get.py

```diff
@@ -1692,9 +1693,25 @@ def package_search(context, data_dict):
         fq = data_dict.get('fq', '')
         if not context.get('ignore_capacity_check', False):
             fq = ' '.join(p for p in fq.split(' ')
-                          if not 'capacity:' in p)
+                          if 'capacity:' not in p)
             data_dict['fq'] = fq + ' capacity:"public"'
 
+        # If present, Solr doesn't need 'include_drafts', so pop it.
+        include_drafts = data_dict.pop('include_drafts', False)
+        fq = data_dict.get('fq', '')
+        if include_drafts:
+            user_id = authz.get_user_id_for_username(user, allow_none=True)
+            if authz.is_sysadmin(user):
+                data_dict['fq'] = fq + ' +state:(active OR draft)'
+            elif user_id:
+                # Query to return all active datasets, and all draft datasets
+                # for this user.
+                data_dict['fq'] = fq + \
+                    ' ((creator_user_id:{0} AND +state:(draft OR active))' \
+                    ' OR state:active)'.format(user_id)
+        elif not authz.is_sysadmin(user):
```

Comment:
If `+state` hasn't been already been set, it will be set in `PackageSearchQuery.run()` to `+state:active`[1], thereby excluding drafts by default. That `elif` on L1712 allows only sysadmins to add a state to their queries explicitly if they want. I've added a test for sysadmins with `include_drafts=False` that passes as expected.

[1] https://github.com/ckan/ckan/blob/b5c43f0ece24f8ff2fd4dd05651a11c1f8e9e23d/ckan/lib/search/query.py#L329